lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php

```diff
@@ -217,18 +234,37 @@ protected function connectTo($connectionName)
     }
 
     /**
-     * @param string $connectionName
-     * @param array  $params
+     * Connets to a random slave and tries others until if found a server that is alive
      *
-     * @return mixed
+     * @param array $params
+     * @param array $driverOptions
+     *
+     * @return Driver\Connection
+     *
+     * @throws ConnectionException
      */
-    protected function chooseConnectionConfiguration($connectionName, $params)
+    private function connectToSlave($params, $driverOptions)
     {
-        if ($connectionName === 'master') {
-            return $params['master'];
+        shuffle($params['slaves']);
+
+        while($connectionParams = array_pop($params['slaves'])) {
+            $user = isset($connectionParams['user']) ? $connectionParams['user'] : null;
+            $password = isset($connectionParams['password']) ? $connectionParams['password'] : null;
+
+            try {
+                return $this->_driver->connect($connectionParams, $user, $password, $driverOptions);
+            } catch (ConnectionException $e) {
+                // try next one
+            }
+        }
+
+        $previous = null;
+
+        if (isset($e) && $e->getPrevious() instanceof \Doctrine\DBAL\Driver\DriverException) {
+            $previous = $e->getPrevious();
         }
 
-        return $params['slaves'][array_rand($params['slaves'])];
+        throw new ConnectionException('No slaves are available', $previous);
```

Comment:
I was not shure about this as well. Thought that it is always better to keep the history for debugging reasons.